src/main/java/org/sqlite/jdbc3/JDBC3DatabaseMetaData.java

```diff
@@ -1011,14 +1015,62 @@ public ResultSet getColumns(String c, String s, String tblNamePattern, String co
                         } else {
                             // catch-all
                             colJavaType = Types.VARCHAR;
                         }
+                        // try to find an (optional) length/dimension of the column
+                        int iStartOfDimension = colType.indexOf('(');
+                        if (iStartOfDimension > 0) {
+                            // find end of dimension
+                            int iEndOfDimension = colType.indexOf(')', iStartOfDimension);
+                            if (iEndOfDimension > 0) {
+                                String sInteger, sDecimal;
+                                // check for two values (integer part, fraction) divided by comma
+                                int iDimensionSeparator = colType.indexOf(',', iStartOfDimension);
+                                if (iDimensionSeparator > 0) {
+                                    sInteger =
+                                            colType.substring(
+                                                    iStartOfDimension + 1, iDimensionSeparator);
+                                    sDecimal =
+                                            colType.substring(
+                                                    iDimensionSeparator + 1, iEndOfDimension);
+                                }
+                                // only a single dimension
+                                else {
+                                    sInteger =
+                                            colType.substring(
+                                                    iStartOfDimension + 1, iEndOfDimension);
+                                    sDecimal = null;
+                                }
+                                // try to parse the values
+                                try {
+                                    int iInteger = Integer.parseUnsignedInt(sInteger);
+                                    // parse decimals?
+                                    if (sDecimal != null) {
+                                        iDecimalDigits = Integer.parseUnsignedInt(sDecimal);
+                                        // columns size equals sum of integer and decimal part of
+                                        // dimension
+                                        iColumnSize = iInteger + iDecimalDigits;
+                                    } else {
+                                        // no decimals
+                                        iDecimalDigits = 0;
+                                        // columns size equals dimension
+                                        iColumnSize = iInteger;
+                                    }
+                                } catch (NumberFormatException ex) {
+                                    // just ignore invalid dimension formats here
+                                }
+                            }
+                        }
```

Comment:
What do you think of using regex instead of `indexOf`/`substring` ?

Would need to define this somewhere:
```java
protected static final Pattern COLUMN_LENGTH = Pattern.compile(".*\\((?<integer>[0-9]+)(,(?<decimal>[0-9]+))?\\)");
```

Then we can replace this whole block:

```suggestion
                        Matcher matcher = COLUMN_LENGTH.matcher(colType);
                        if (matcher.matches()) {
                            try {
                                iDecimalDigits = matcher.group("decimal") == null ? 0 : Integer.parseUnsignedInt(matcher.group("decimal"));
                                iColumnSize = Integer.parseUnsignedInt(matcher.group("integer")) + iDecimalDigits;
                            } catch (NumberFormatException ignored) {
                            }
                        }
```

Reply:
Your welcome :smile:

I would not use a regex as it is much slower than using a plain `indexOf` (which is just a small loop under the hood).
And the parsing calls would still be necessary.

Reply:
> I would not use a regex as it is much slower than using a plain `indexOf`

Is it? Even when the pattern is compiled only once?

Reply:
Yes, it is. There is no regex evaluation, which is much more expensive than just looping through a string by characters.
Though you could replace it, if you really like to and want more concise code.
I just wanted to add very performant code with a low memory overhead (small number of objects constructed on the heap). As I thought this might be called pretty often if multiple tables are queried or tables with a lot of columns.

I would use regex for more complex expressions, but here it is just to make sure two brackets are there and distinguishing between one and two parameters. Both variants still have to parse the value(s) to integers, so that part is the same.